resources/testdriver.js

```diff
@@ -0,0 +1,84 @@
+(function() {
+    "use strict";
+
+    function getInViewCenterPoint(rect) {
+        var left = Math.max(0, Math.min(rect.x, rect.x + rect.width));
+        var right = Math.min(window.innerWidth, Math.max(rect.x, rect.x + rect.width));
+        var top = Math.max(0, Math.min(rect.y, rect.y + rect.height));
+        var bottom = Math.min(window.innerHeight, Math.max(rect.y, rect.y + rect.height));
```

Comment:
FYI: innerWidth and innerHeight are layout viewport in Chrome only (well, Firefox technically too since both viewports are the same). In Safari and Edge, there isn't really a good way to get the layout viewport size...you can add an invisible position: fixed Element that's `width: 100%;  height: 100%` and read its `getBoundingClientRect` but that interferes with the content. You could read the getBoundingClientRect of the `document.documentElement` which works in most cases but has some edge cases that it falls down on (explicitly sized `documentElement`, weird meta tags, etc.).